File: WayFinder/ViewModels/OutfitWeatherViewModel.swift
import Foundation
import SwiftUI
import UIKit

@MainActor
class OutfitWeatherViewModel: ObservableObject {
    @Published var uiState: OutfitWeatherUiState = .idle
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var outfitHistory: [Outfit] = []
    @Published var isLoadingHistory: Bool = false
    
    private let outfitService = OutfitWeatherService.shared
    
    var outfit: Outfit? {
        if case .success(let outfit) = uiState {
            return outfit
        }
        return nil
    }
    
    func uploadOutfit(imageData: Data, bookingId: String) async {
        isLoading = true
        errorMessage = nil
        uiState = .loading
        
        do {
            let response = try await outfitService.uploadOutfitImage(imageData: imageData, bookingId: bookingId)
            uiState = .success(response.analysis)
            isLoading = false
        } catch {
            var errorMsg = error.localizedDescription
            
            // Améliorer le message d'erreur pour les timeouts
            if let urlError = error as? URLError {
                switch urlError.code {
                case .timedOut:
                    errorMsg = String(localized: "outfit_timeout_message")
                case .notConnectedToInternet:
                    errorMsg = String(localized: "outfit_no_internet")
                case .cannotConnectToHost:
                    errorMsg = String(localized: "outfit_server_error")
                default:
                    break
                }
            }
            
            errorMessage = errorMsg
            uiState = .error(errorMsg)
            isLoading = false
        }
    }
    
    func loadOutfit(outfitId: String) async {
        isLoading = true
        errorMessage = nil
        uiState = .loading
        
        do {
            let outfit = try await outfitService.getOutfit(outfitId: outfitId)
            uiState = .success(outfit)
            isLoading = false
        } catch {
            let errorMsg = error.localizedDescription
            errorMessage = errorMsg
            uiState = .error(errorMsg)
            isLoading = false
        }
    }
    
    func loadOutfitHistory(bookingId: String) async {
        isLoadingHistory = true
        do {
            let outfits = try await outfitService.getOutfitsForBooking(bookingId: bookingId)
            outfitHistory = outfits.sorted { outfit1, outfit2 in
                // Sort by outfit_date if available, otherwise by createdAt
                if let d1 = outfit1.outfitDate, let d2 = outfit2.outfitDate {
                    return d1 > d2
                }
                if outfit1.outfitDate != nil {
                    return true
                }
                if outfit2.outfitDate != nil {
                    return false
                }
                return (outfit1.createdAt ?? "") > (outfit2.createdAt ?? "")
            }
            isLoadingHistory = false
        } catch {
            print("❌ [OutfitWeatherViewModel] Error loading history: \(error)")
            isLoadingHistory = false
        }
    }
    
    func deleteOutfit(outfitId: String, bookingId: String) async {
        do {
            try await outfitService.deleteOutfit(outfitId: outfitId)
            // Recharger l'historique après suppression
            await loadOutfitHistory(bookingId: bookingId)
        } catch {
            print("❌ [OutfitWeatherViewModel] Error deleting outfit: \(error)")
            errorMessage = error.localizedDescription
        }
    }
    
    func resetState() {
        uiState = .idle
        errorMessage = nil
        isLoading = false
    }
}

enum OutfitWeatherUiState: Equatable {
    case idle
    case loading
    case success(Outfit)
    case error(String)
    
    static func == (lhs: OutfitWeatherUiState, rhs: OutfitWeatherUiState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle), (.loading, .loading):
            return true
        case (.success(let lhsOutfit), .success(let rhsOutfit)):
            return lhsOutfit.id == rhsOutfit.id
        case (.error(let lhsError), .error(let rhsError)):
            return lhsError == rhsError
        default:
            return false
        }
    }
}

File: app/src/main/java/tn/esprit/wayfinder/viewmodels/OutfitWeatherViewModel.kt
package tn.esprit.wayfinder.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import retrofit2.HttpException
import tn.esprit.wayfinder.models.Outfit
import tn.esprit.wayfinder.network.ApiService
import java.io.IOException

sealed class OutfitWeatherUiState {
    object Idle : OutfitWeatherUiState()
    object Loading : OutfitWeatherUiState()
    data class Success(val outfit: Outfit) : OutfitWeatherUiState()
    data class Error(val message: String) : OutfitWeatherUiState()
}

class OutfitWeatherViewModel(private val apiService: ApiService) : ViewModel() {

    private val _uiState = MutableStateFlow<OutfitWeatherUiState>(OutfitWeatherUiState.Idle)
    val uiState: StateFlow<OutfitWeatherUiState> = _uiState.asStateFlow()

    private val _outfitHistory = MutableStateFlow<List<Outfit>>(emptyList())
    val outfitHistory: StateFlow<List<Outfit>> = _outfitHistory.asStateFlow()

    private val _isLoadingHistory = MutableStateFlow(false)
    val isLoadingHistory: StateFlow<Boolean> = _isLoadingHistory.asStateFlow()

    val isLoading: Boolean
        get() = _uiState.value is OutfitWeatherUiState.Loading

    val outfit: Outfit?
        get() = (_uiState.value as? OutfitWeatherUiState.Success)?.outfit

    val errorMessage: String?
        get() = (_uiState.value as? OutfitWeatherUiState.Error)?.message

    fun uploadOutfit(imagePart: MultipartBody.Part, bookingId: String) {
        viewModelScope.launch {
            _uiState.value = OutfitWeatherUiState.Loading
            try {
                android.util.Log.d("OutfitWeatherViewModel", "Starting upload for booking: $bookingId")
                val bookingIdBody = bookingId.toRequestBody("text/plain".toMediaType())
                val response = apiService.uploadOutfitImage(imagePart, bookingIdBody)
                android.util.Log.d("OutfitWeatherViewModel", "Upload successful, analysis: ${response.analysis}")
                _uiState.value = OutfitWeatherUiState.Success(response.analysis)
            } catch (e: HttpException) {
                val errorMessage = try {
                    val errorBody = e.response()?.errorBody()?.string()
                    android.util.Log.e("OutfitWeatherViewModel", "HTTP error: ${e.code()}, body: $errorBody")
                    errorBody ?: "Erreur lors de l'upload (${e.code()})"
                } catch (ex: Exception) {
                    android.util.Log.e("OutfitWeatherViewModel", "Error reading error body: ${ex.message}")
                    "Erreur lors de l'upload de l'image"
                }
                _uiState.value = OutfitWeatherUiState.Error(errorMessage)
            } catch (e: IOException) {
                android.util.Log.e("OutfitWeatherViewModel", "IO error: ${e.message}")
                _uiState.value = OutfitWeatherUiState.Error("Erreur de connexion. Vérifiez votre Internet.")
            } catch (e: Exception) {
                android.util.Log.e("OutfitWeatherViewModel", "Unexpected error: ${e.message}", e)
                _uiState.value = OutfitWeatherUiState.Error("Erreur inattendue: ${e.message ?: "Erreur inconnue"}")
            }
        }
    }

    fun analyzeOutfit(bookingId: String, imageUrl: String) {
        viewModelScope.launch {
            _uiState.value = OutfitWeatherUiState.Loading
            try {
                val request = tn.esprit.wayfinder.models.AnalyzeOutfitRequest(bookingId, imageUrl)
                val outfit = apiService.analyzeOutfit(request)
                _uiState.value = OutfitWeatherUiState.Success(outfit)
            } catch (e: HttpException) {
                val errorMessage = try {
                    e.response()?.errorBody()?.string() ?: "Erreur lors de l'analyse"
                } catch (ex: Exception) {
                    "Erreur lors de l'analyse de la tenue"
                }
                _uiState.value = OutfitWeatherUiState.Error(errorMessage)
            } catch (e: IOException) {
                _uiState.value = OutfitWeatherUiState.Error("Erreur de connexion. Vérifiez votre Internet.")
            } catch (e: Exception) {
                _uiState.value = OutfitWeatherUiState.Error("Erreur inattendue: ${e.message}")
            }
        }
    }

    fun loadOutfit(outfitId: String) {
        viewModelScope.launch {
            _uiState.value = OutfitWeatherUiState.Loading
            try {
                android.util.Log.d("OutfitWeatherViewModel", "Loading outfit: $outfitId")
                val outfit = apiService.getOutfit(outfitId)
                android.util.Log.d("OutfitWeatherViewModel", "Outfit loaded successfully: ${outfit._id}")
                _uiState.value = OutfitWeatherUiState.Success(outfit)
            } catch (e: HttpException) {
                val errorMessage = try {
                    val errorBody = e.response()?.errorBody()?.string()
                    android.util.Log.e("OutfitWeatherViewModel", "HTTP error: ${e.code()}, body: $errorBody")
                    errorBody ?: "Erreur lors du chargement (${e.code()})"
                } catch (ex: Exception) {
                    android.util.Log.e("OutfitWeatherViewModel", "Error reading error body: ${ex.message}")
                    "Erreur lors du chargement de l'outfit"
                }
                _uiState.value = OutfitWeatherUiState.Error(errorMessage)
            } catch (e: IOException) {
                android.util.Log.e("OutfitWeatherViewModel", "IO error: ${e.message}")
                _uiState.value = OutfitWeatherUiState.Error("Erreur de connexion. Vérifiez votre Internet.")
            } catch (e: Exception) {
                android.util.Log.e("OutfitWeatherViewModel", "Unexpected error: ${e.message}", e)
                _uiState.value = OutfitWeatherUiState.Error("Erreur inattendue: ${e.message ?: "Erreur inconnue"}")
            }
        }
    }

    fun loadOutfitHistory(bookingId: String) {
        viewModelScope.launch {
            _isLoadingHistory.value = true
            try {
                android.util.Log.d("OutfitWeatherViewModel", "Loading outfit history for booking: $bookingId")
                val outfits = apiService.getOutfitsForBooking(bookingId)
                android.util.Log.d("OutfitWeatherViewModel", "Loaded ${outfits.size} outfits")
                _outfitHistory.value = outfits.sortedByDescending { it.createdAt }
            } catch (e: HttpException) {
                android.util.Log.e("OutfitWeatherViewModel", "HTTP error loading history: ${e.code()}")
                _outfitHistory.value = emptyList()
            } catch (e: IOException) {
                android.util.Log.e("OutfitWeatherViewModel", "IO error loading history: ${e.message}")
                _outfitHistory.value = emptyList()
            } catch (e: Exception) {
                android.util.Log.e("OutfitWeatherViewModel", "Unexpected error loading history: ${e.message}", e)
                _outfitHistory.value = emptyList()
            } finally {
                _isLoadingHistory.value = false
            }
        }
    }

    fun deleteOutfit(outfitId: String, bookingId: String) {
        viewModelScope.launch {
            try {
                android.util.Log.d("OutfitWeatherViewModel", "Deleting outfit: $outfitId for booking: $bookingId")
                apiService.deleteOutfit(outfitId)
                android.util.Log.d("OutfitWeatherViewModel", "Outfit deleted successfully, refreshing history")
                // Remove from local state immediately for better UX
                _outfitHistory.value = _outfitHistory.value.filter { it._id != outfitId }
                // Refresh history from server to ensure consistency
                loadOutfitHistory(bookingId)
            } catch (e: HttpException) {
                android.util.Log.e("OutfitWeatherViewModel", "HTTP error deleting outfit: ${e.code()}, ${e.message()}")
                // Reload history even on error to show current state
                loadOutfitHistory(bookingId)
            } catch (e: IOException) {
                android.util.Log.e("OutfitWeatherViewModel", "IO error deleting outfit: ${e.message}")
                // Reload history even on error to show current state
                loadOutfitHistory(bookingId)
            } catch (e: Exception) {
                android.util.Log.e("OutfitWeatherViewModel", "Unexpected error deleting outfit: ${e.message}", e)
                // Reload history even on error to show current state
                loadOutfitHistory(bookingId)
            }
        }
    }

    fun resetState() {
        _uiState.value = OutfitWeatherUiState.Idle
    }
}

